owid/etl
# :tada: Updates to processing log

Update processing log to handle more cases like renamed variables, etc. It adds a new field `target` that is similar to `variable_name`, but with an additional random suffix that lets us reconstruct the complete computational graph.

I refactored processing log a bit by:
* Creating ProcessingLog and LogEntry objects
* Removed `amend_log` (I didn't find any use for it and thought it'd be nicer if processing log was immutable)
* Moved the logic to `processing_log.add_entry` and `variable.update_log` with sensible defaults

I got it into a state where it's mergeable, though definitely not perfect. I've only polished it on a few examples. There's a lot to do in terms of refactoring and testing, but I think it only makes sense to go all in once we confirm this is useful.

## How to review:

I'd recommend creating a new `etl-plog` directory next to `etl` and experimenting with the processing log there, otherwise you end up with a mix of plog / non-plog datasets which could be confusing (they work together, but it's still cleaner).

1. Check out docs at `docs/architecture/metadata/processing_log.md`

1. Delete all datasets that used old processing log
```
find data/ -name "*.meta.json" -exec grep -l 'processing_log' {} + | xargs -I{} grep -L 'target' {} | xargs -I{} dirname {} | sort -u | xargs -I{} rm -rf {}
```
2. Rerun dataset with processing log enabled
```
PROCESSING_LOG=1 etl uk_egg_statistics --force
```
3. In notebook, visualise the log
```
from owid.catalog import Dataset
from etl.paths import DATA_DIR

ds = Dataset(DATA_DIR / "garden/animal_welfare/2023-08-01/uk_egg_statistics")
tab = ds['uk_egg_statistics']
tab['number_of_eggs_all'].m.processing_log.display(DATA_DIR)
```

## Open questions

To enable plog you have to run it with `PROCESSING_LOG=1 etl ...`, but you cannot simply add it to your `.env` file, because it'd ruin displaying plog from notebooks. It's not as great, but I found no easier way. The nice thing about this is exploring plog for your datasets. You can just run
```
PROCESSING_LOG=1 etl ... --force --only
```
and then check it from the notebook.

## Examples

Here's an artificial example. I've only tried a couple of datasets.

<img width="904" alt="image" src="https://github.com/owid/etl/assets/1550888/89689f63-caca-4eb8-bea8-12d77bd1e895">
